smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -301,7 +495,56 @@ protected abstract void serializeDocument(
             List<HttpBinding> documentBindings
     );
 
-    private void generateOperationDeserializer(GenerationContext context, OperationShape operation, HttpTrait trait) {
+    /**
+     * Writes the code needed to serialize a structure in the document of a request.
+     *
+     * @param context The generation context.
+     * @param shape The structured shape being generated.
+     */
+    protected abstract void serializeDocumentStructure(
```

Comment:
My intention for the `serializeDocument` method was that it would be a black box for handling how all input documents are serialized. How is it being used now? The same way? If so, how is stuff like `serializeDocumentUnion` different?

Reply:
It's the same way (after I move a line back that I forgot to in R1.) The `serializeDocument[Shape]` functions handle the same black box of content, but for the methods that serialize for other document shapes instead of the main document shape.

```ts
// Validate we have input, then get the right serialization for the member target.
writer.write("input &&");
writer.openBlock("input.map(entry =>", ");", () -> {
    writer.write(getInputValue(context, Location.DOCUMENT, shape, shape.getMember(), target));
});
```

However, where the structure function is invoked could be a touch more similar to the current `serializeDocument`.